pinot-core/src/main/java/com/linkedin/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -33,318 +36,266 @@
- * to BrokerResponseNative.
- *
+ * The <code>BrokerReduceService</code> class provides service to reduce data tables gathered from multiple servers
+ * to {@link BrokerResponseNative}.
  */
 public class BrokerReduceService implements ReduceService<BrokerResponseNative> {
   private static final Logger LOGGER = LoggerFactory.getLogger(BrokerReduceService.class);
 
   private static final String NUM_DOCS_SCANNED = "numDocsScanned";
   private static final String NUM_ENTRIES_SCANNED_IN_FILTER = "numEntriesScannedInFilter";
-  private static final String NUM_ENTREIS_SCANNED_POST_FILTER = "numEntriesScannedPostFilter";
+  private static final String NUM_ENTRIES_SCANNED_POST_FILTER = "numEntriesScannedPostFilter";
   private static final String TOTAL_DOCS = "totalDocs";
 
   // Not used currently, but it records the time spent on server side.
   private static final String TIME_USED_MS = "timeUsedMs";
 
   @Override
-  public BrokerResponseNative reduceOnDataTable(BrokerRequest brokerRequest,
-      Map<ServerInstance, DataTable> dataTableMap) {
-    // Empty result.
-    if (dataTableMap == null || dataTableMap.size() == 0) {
-      return BrokerResponseNative.EMPTY_RESULT;
+  public BrokerResponseNative reduceOnDataTable(@Nonnull BrokerRequest brokerRequest,
+      @Nonnull Map<ServerInstance, DataTable> dataTableMap) {
+    BrokerResponseNative brokerResponseNative = new BrokerResponseNative();
+
+    if (dataTableMap.size() == 0) {
+      // Empty response.
+      return brokerResponseNative;
     }
 
-    BrokerResponseNative brokerResponseNative = new BrokerResponseNative();
+    // Process server response metadata.
+    List<QueryProcessingException> processingExceptions = brokerResponseNative.getProcessingExceptions();
     long numDocsScanned = 0L;
     long numEntriesScannedInFilter = 0L;
     long numEntriesScannedPostFilter = 0L;
     long numTotalRawDocs = 0L;
-
     Iterator<Map.Entry<ServerInstance, DataTable>> iterator = dataTableMap.entrySet().iterator();
+    Map.Entry<ServerInstance, DataTable> entryWithDataSchema = null;
     while (iterator.hasNext()) {
       Map.Entry<ServerInstance, DataTable> entry = iterator.next();
       ServerInstance serverInstance = entry.getKey();
-      DataTable dataTable = dataTableMap.get(serverInstance);
-      if (dataTable == null) {
-        continue;
-      }
+      DataTable dataTable = entry.getValue();
+      Map<String, String> metadata = dataTable.getMetadata();
 
-      // reduceOnTraceInfo (put it here so that trace info can show up even exception happens)
-      if (brokerRequest.isEnableTrace() && dataTable.getMetadata() != null) {
-        brokerResponseNative.getTraceInfo()
-            .put(serverInstance.getHostname(), dataTable.getMetadata().get("traceInfo"));
+      // Reduce on trace info.
+      if (brokerRequest.isEnableTrace()) {
+        brokerResponseNative.getTraceInfo().put(serverInstance.getHostname(), metadata.get("traceInfo"));
```

Comment:
will be good to have all metadata key constants defined in datatable. May be create an enum